app/src/crypto.c

```diff
@@ -22,108 +22,87 @@
 
 uint32_t hdPath[HDPATH_LEN_DEFAULT];
 
-zxerr_t crypto_extractPublicKey(uint8_t *pubKey, uint16_t pubKeyLen)
-{
+zxerr_t crypto_extractPublicKey(uint8_t *pubKey, uint16_t pubKeyLen) {
+    if (pubKey == NULL || pubKeyLen < PK_LEN_25519) {
+        return zxerr_invalid_crypto_settings;
+    }
+
+    zxerr_t error = zxerr_unknown;
     cx_ecfp_public_key_t cx_publicKey;
     cx_ecfp_private_key_t cx_privateKey;
-    uint8_t privateKeyData[SK_LEN_25519];
+    uint8_t privateKeyData[SK_LEN_25519] = {0};
 
-    if (pubKeyLen < PK_LEN_25519) {
-        return zxerr_invalid_crypto_settings;
+    // Generate keys
+    CATCH_CXERROR(os_derive_bip32_with_seed_no_throw(HDW_NORMAL,
+                                                     CX_CURVE_Ed25519,
+                                                     hdPath,
+                                                     HDPATH_LEN_DEFAULT,
+                                                     privateKeyData,
+                                                     NULL,
+                                                     NULL,
+                                                     0))
+
+    CATCH_CXERROR(cx_ecfp_init_private_key_no_throw(CX_CURVE_Ed25519, privateKeyData, 32, &cx_privateKey))
+    CATCH_CXERROR(cx_ecfp_init_public_key_no_throw(CX_CURVE_Ed25519, NULL, 0, &cx_publicKey))
+    CATCH_CXERROR(cx_ecfp_generate_pair_no_throw(CX_CURVE_Ed25519, &cx_publicKey, &cx_privateKey, 1))
```

Comment:
@ftheirs Could be replaced in future version with a single call to https://github.com/LedgerHQ/ledger-secure-sdk/blob/master/lib_standard_app/crypto_helpers.h#L133
that also handle the memset of private key data.